tools/extract_cmake_functions.py

```diff
@@ -0,0 +1,137 @@
+# \brief\ <brief description, which should be included in the main description>
+# \brief_nodesc\ <brief description, which should be excluded from the main description>
+# \param\ <parameter_name> <parameter description>
+#
+# Commands in the middle of a line are not supported
+def process_commands(description, name, group, signature):
+    brief = ''
+    params = []
+    ret = ''
+    for line in description.split('\n'):
+        line = line.strip()
+        if line.startswith('\\'):
+            command = line.split('\\')[1]
+            if command == 'param':
+                # Parameter name and description
+                params.append(line.split('\\')[2].strip())
+            elif command == 'brief':
+                # Brief description
+                brief = line.split('\\')[2].strip()
+                ret += brief + '\n'
+            elif command == 'brief_nodesc':
+                # Brief description which should not be included in the main description
+                brief = line.split('\\')[2].strip()
+            else:
+                logger.error("{}:{} has unknown command: {}".format(group, name, command))
+        elif '\\' in line:
+            logger.error("{}:{} has a line containing '\\': {}".format(group, name, line))
+        else:
+            ret += line + '\n'
+    # Check that there are no semicolons in the parameter descriptions, as that's the delimiter for the parameter list
+    if any([';' in x for x in params]):
+        logger.error("{}:{} has a parameter description containing ';'".format(group, name))
+    # Check that all parameters are in the signature
+    for param in params:
+        if param.split(' ')[0] not in signature:
+            logger.error("{}:{} has a parameter {} which is not in the signature {}".format(group, name, param.split(' ')[0], signature))
+    # Check that the brief description is not empty
+    if not brief:
+        logger.warning("{}:{} has no brief description".format(group, name))
+
+    return ret.strip(), brief, ';'.join(params)
+
+
+# Scan all CMakeLists.txt and .cmake files in the specific path, recursively.
+
+for dirpath, dirnames, filenames in os.walk(scandir):
+    for filename in filenames:
+        if filename in skip_files:
+            continue
+        group = os.path.basename(dirpath)
+        if group in skip_groups:
+            continue
+        file_ext = os.path.splitext(filename)[1]
+        if filename == 'CMakeLists.txt' or file_ext == '.cmake':
+            file_path = os.path.join(dirpath, filename)
+
+            with open(file_path, encoding="ISO-8859-1") as fh:
+                text = fh.read()
+                for match in CMAKE_FUNCTION_RE.finditer(text):
+                    name = match.group(4)
+                    signature = match.group(1).strip()
+                    if signature.startswith(name):
+                        description, brief, params = process_commands(match.group(2).replace('#', ''), name, group, signature)
```

Comment:
One of the descriptions has a genuine `#` inside it (`# To add additional directories to the gatt #import path, add them to the end of the argument list.`), so IMHO the `.replace('#', '')` here ought to be moved inside the line-by-line processing in `process_commands` and then you could do something like `line = re.sub('^\s+#', '', line)` to only remove the `#` from the start of lines.

Reply:
I think it's simpler to just not support `#` in the descriptions, so that would make `#`, `;` and `\` unsupported/special characters (I'll add a note in the script) - I'll remove that one, as it's not really needed